pkg/modifycontroller/controller.go

```diff
@@ -0,0 +1,349 @@
+	oldPVC, ok := oldObj.(*v1.PersistentVolumeClaim)
+	if !ok || oldPVC == nil {
+		return
+	}
+
+	newPVC, ok := newObj.(*v1.PersistentVolumeClaim)
+	if !ok || newPVC == nil {
+		return
+	}
+
+	// Only trigger modify volume if the following conditions are met
+	// 1. Non empty vac name
+	// 2. oldVacName != newVacName
+	// 3. PVC is in Bound state
+	oldVacName := oldPVC.Spec.VolumeAttributesClassName
+	newVacName := newPVC.Spec.VolumeAttributesClassName
+	if newVacName != nil && *newVacName != "" && *newVacName != *oldVacName && oldPVC.Status.Phase == v1.ClaimBound {
+		volumeObj, exists, err := ctrl.volumes.GetByKey(oldPVC.Spec.VolumeName)
+		if err != nil {
+			klog.Errorf("Get PV %q of pvc %q failed: %v", oldPVC.Spec.VolumeName, klog.KObj(oldPVC), err)
+			return
+		}
+		if !exists {
+			klog.InfoS("PV bound to PVC not found", "PV", oldPVC.Spec.VolumeName, "PVC", klog.KObj(oldPVC))
+			return
+		}
+
+		_, ok := volumeObj.(*v1.PersistentVolume)
+		if !ok {
+			klog.Errorf("expected volume but got %+v", volumeObj)
+			return
+		}
+
+		// Handle modify volume by adding to the claimQueue to avoid race conditions
+		ctrl.addPVC(newObj)
+	} else {
+		klog.V(4).InfoS("No need to modify PVC", "PVC", klog.KObj(newPVC))
+	}
+}
+
+func (ctrl *modifyController) deletePVC(obj interface{}) {
+	objKey, err := util.GetObjectKey(obj)
+	if err != nil {
+		return
+	}
+	ctrl.claimQueue.Forget(objKey)
+}
+
+func (ctrl *modifyController) addVAC(obj interface{}) {
+	objKey, err := util.GetObjectKey(obj)
+	if err != nil {
+		return
+	}
+	ctrl.vacQueue.Add(objKey)
+}
+
+func (ctrl *modifyController) deleteVAC(obj interface{}) {
+	objKey, err := util.GetObjectKey(obj)
+	if err != nil {
+		return
+	}
+	ctrl.vacQueue.Forget(objKey)
+}
+
+// modifyPVC modifies the PVC and PV based on VAC
+func (ctrl *modifyController) modifyPVC(pvc *v1.PersistentVolumeClaim, pv *v1.PersistentVolume) error {
+	var err error
+	if isFirstTimeModifyVolumeWithPVC(pvc, pv) {
+		// If it is first time adding a vac, always validate and then call modify volume
+		_, _, err, _ = ctrl.validateVACAndModifyVolumeWithTarget(pvc, pv)
+	} else {
+		_, _, err, _ = ctrl.modify(pvc, pv)
+	}
+	return err
+}
+
+func isFirstTimeModifyVolumeWithPVC(pvc *v1.PersistentVolumeClaim, pv *v1.PersistentVolume) bool {
+	if pv.Spec.VolumeAttributesClassName == nil && pvc.Spec.VolumeAttributesClassName != nil {
+		return true
+	}
+	return false
+}
+
+// Run starts the controller.
+func (ctrl *modifyController) Run(
+	workers int, ctx context.Context) {
+	defer ctrl.claimQueue.ShutDown()
+
+	klog.InfoS("Starting external resizer for modify volume", "controller", ctrl.name)
+	defer klog.InfoS("Shutting down external resizer", "controller", ctrl.name)
+
+	stopCh := ctx.Done()
+	informersSyncd := []cache.InformerSynced{ctrl.pvSynced, ctrl.pvcSynced}
+
+	// Only WaitForCacheSync for VAC if feature gate is enabled
+	if utilfeature.DefaultFeatureGate.Enabled(features.VolumeAttributesClass) {
```

Comment:
nit: Do we even need these checks here? We should not even be running this controller if `modifyVolume` feature is not enabled.